File: BinaryTree/JAVA/src/implementing_BT/BalancedBT.java
package implementing_BT;

public class BalancedBT {
	Node root;

	public int isBalanced(Node node) {
		if (node == null) {
			return 0;
		}
		int leftSubTreeHeight = isBalanced(node.left); // recursive call basically until node == null
		if (leftSubTreeHeight == -1)
			return -1;
		int rightSubTreeHeight = isBalanced(node.right);
		if (rightSubTreeHeight == -1)
			return -1;

		if (Math.abs(leftSubTreeHeight - rightSubTreeHeight) > 1) // difference in height of left\right subtrees
			return -1;

		return (Math.max(leftSubTreeHeight, rightSubTreeHeight) + 1);
	}

	public static void main(String[] args) {
		BalancedBT myTree = new BalancedBT();
		myTree.root = new Node(10);
		myTree.root.left = new Node(8);
		myTree.root.right = new Node(11);
		myTree.root.left.left = new Node(7);
		myTree.root.left.right = new Node(9);
		// myTree.root.left.left.left = new Node(5);

		if (myTree.isBalanced(myTree.root) != -1)
			System.out.println("This is a Balanced BT!");
		else
			System.out.println("This is not a Balanced BT!");
	}
}

File: BinaryTree/Python/BalancedBT.py
class Node:
    def __init__(self, data):
        self.data = data
        self.right = None
        self.left = None


def is_balanced(node):
    if node is None:
        return 0
    left_subtree_height = is_balanced(node.left)
    if left_subtree_height == -1:
        return -1
    right_subtree_height = is_balanced(node.right)
    if right_subtree_height == -1:
        return -1

    if abs(left_subtree_height - right_subtree_height) > 1:
        return -1

    return max(left_subtree_height, right_subtree_height) + 1


root = Node(10)
root.left = Node(8)
root.right = Node(11)
root.left.left = Node(7)
root.left.right = Node(9)
# root.left.left.left = Node(5)

if is_balanced(root) != -1:
    print("This is a Balanced BT!")
else:
    print("This is not a Balanced BT!")
